README
Name: Won Seok Yang
Class: Functional Programming in Ocaml
Date: June 26, 2019

====================
To run the program:
  "make"
  "./game"
====================


Project
For this class I will be making a turnbased RTS using Ocaml and it's builtin
Graphics module. I will also be learning how to use github!

I want to create a game where the player is presented with a gameboard of 
units they control and units the computer controls. The objective of the 
game is to eliminte all enemy units using the units provided.
There will be different unit types with different stats as well as different
terrain types that will affect gameplay.


I will list goals/milestones below and add/check them off as I progress.
For a more detailed history check "changelog.txt".

I will first start by creating basic base elements and expanding upon it as
the program gets built out.

Ultimately, I want the program to fit a few criteria:
	[ ] 3-5 unit types               {HP, ATK, RNG, MV}
	  [-] Melee: basic attack unt    { 3, 1, 1, 1}
	  [ ] Range: increase atk range  { 2, 1, 2, 1}
	  [ ] Calvery: increase MV       { 2, 1, 1, 3}
	[X] User cursor to select grid and units
	[ ] Implimentation of simple move and attack AI
	[-] Different terrain types

If I have enough time I'd like to accomplish:	
	[ ] A menu
	  [ ] Variable playable grid size
	[ ] An info panel displaying terrain and unit info of current tile to player  
	[ ] A pleasing GUI
	[ ] Different terrain effects
	  [ ] Forest: decrease attack range by 1
	  [ ] Mountain/Hills: decrease movement by 1 but increase range by 1
	  [ ] Water: If entered this turn, unit cannot attack

Thing's that would be awesome but I probably won't be able to implement in time:
	[ ] Individual unit leveling
	  [ ] Increases in stats per level
	  [ ] Class upgrades with enough levels giving major buffs 
	[ ] Fog of war that hides parts of the map unless a player unit has occupied it
	[ ] Different levels/grid sets for the player to go through	
	

Initial Steps:
[-] Create a 10 x 10 grid to display game elements
  [X] A data structure to keep track of the grid and it's contents
    [X] Basic terrain png
  [X] Implemented 2D list to gameboard graphics

[ ] Create a unit structure
  [-] Create first unit
    [X] Create first unit sprite

[~] Learn how to use github

[X] Create basic cursor
  [X] Implement user input
  [X] Implement screen boundaries
  [X] Implement select and cancel 

[ ] Fixing input lag (caused by constant png to image conversion)



changelog.txt
I figured I should keep a more detailed log of my changes to make it easier
for the professor to see what changed and so I myself can see the progress.

7/1/19 1:27AM
+I got a rudementary version of the cursor implemented. Which includes drawing
the screen and getting user input "real time". 

To make this better I can stagger the drawing of the screen. Technically, 
the image doesn't need to change until the user has input something 
(unless I later want to implement idle animation for the units :D).

If I'm doing this with only the Graphics module I have to layer the drawing 
of the screen. The BG obviously goes on the bottom (terrain), then the unit,
then the cursor with a transparent middle (only draw the edges the 
cursor is highlighting).

As for the hud with information I'd like to implement, maybe I can make the 
screen a bit larger (by one grid), fill it with a neutral color and put text
over it for the info about unit/terrain.

I thought about using an image editor to draw the BG image png and importing
it to the screen but that would mean the tiles would be set on loading.
Instead, I think I should draw tile types (forest, ground, mountain/hill)
and load them per tile on the grid. I think I can compile them all once, save
it, and redraw when the screen updates.

Since I'm at it I think I'll make the cursor just a border and commit the files.

7/3/19 6:32PM
+Added the boundry box so the users cursor can't be out of bounds. This
change was made possible by adding a (w,h) param to the push function.

7:39PM
+Made the 2D list to hold tile types and drew a single tile for testing. Now
I have to make the position variable depending on the board list and the tile
base color reflect the tile type. I'm commiting this now because my battery
bank is getting low on power and I'm at a cafe.. I would hate to lose any 
progress lol

8:49PM
+Drew the actual game tiles by taking a 2D list of terrain types. 
As of now, the gameboard is being drawn every state update but it doesn't 
have to be. It can be initialized once into a single image with blit_image
maybe?? Anyways, I'm just super stoked it worked like the pseudocode from my
notebook!

7/7/19 11:28PM
I thought I had another week! Turns out tomorrow is the project presentation.. 
FML I even requested a few extra days off next week so I can focus on the project..
I should have paid attention to the due dates. ANYWAYS, now I'm trying to get a place holder
for unit sprites to work. I thought it would be as simple as plugging them into imagelib
and graphics but turns out size is a big problem.

7/9/19 4:57AM
+Added in the Select and Cancel type for user input.

I was laying in bed and my mind kept running with ideas I wanted to try so I
got up and decided to run with it. I initally wanted to make a record type
for the units but then started thinking about how they would integrate with
the game grid. Maybe I could change the list type of the grid to also hold
unit type? Currently the gameboard is a 2D list of type terrain. If I change
the list type to (terrain * unit) then I can draw the terrain and units on
the screen at once..

But then another problem arises. How about the relationship between the 
cursor and the terrain/units? How would I translate the cursor position to
the grid? 

Here are some thoughts:
  The cursor is a state of (x, y) coordinates. So if the cursor is at (0, 0)
I know that it's highlighting the first element in the first list of terrain
list or [ [0], _ ]. If the cursor is at (1unit, 0) of the grid, it's at 
[ [1], _] of the board. 
  So the ypos of the cursor is the n'th list of the list and the xpos is 
the m'th element of the n'th list. If the list of terrain type was stored
in an array that can be reffered to by index this would make selecting a 
point on the board a whole lot easier..
  I vaguely remember seeing an Array module somewhere.. Time to google!
  So it turns out there is an Array module! :D but I've never used it before.
If I want to use an array in the program, I'll have to make a bunch of 
changes.. Let's test the Array module on utop before messing with the main
program.

7/10/19 12:37AM
  So I looked into the Array module and played around with it in utop.
Addressing an array matrix looks easy enough but I kinda want to try it with
the list first.
  Also, I figured out where I went wrong with loading the terrain png! I was
able to get the mountain png to load.

7/11/19 2:00PM
  After I implemented the forest png I noticed that the program was very slow.
I think this is due to the fact that the whole background has to get processed at every screen update. I can fix this by either saving the processed
pgn tiles into a Graphics image type or only letting the program run on user 
input. It would be best to do both and I'd like to do both but for I'll start
with the former.
  The plan is to have the game board initialized once at the start of the 
program and to save the processed "screencap" saved into board_image. This
way, when the screen has to be redrawn I can just use board_image instead of
reprocessing tiles over and over.
  Lets give it a try!

7/13/19 4:44PM
+Program only processes terrain png and builds the board just once at the
start of the game. Every subsequent draw is taken from the Graphics image
type that's been saved

7/16/19 1:52AM
  I was trying to neaten the code further (restructure the code to use
a coordinate system to make implementing units and their position easier)
and I ran into a problem. I use the Graphics module and some of it's 
untions in my draw_screen file but I'm getting a "Fatal Error" because
Graphics window has not been opened.. but it has in the main file. I think
this is because the Graphics is opened in the main "g_test" file and the
compiler doesn't link the call to Graphics in "draw_screen" to the "g_test"
.. To solve this, I could move ALL of the graphics operations to the 
"draw_screen" file.. I think..



g_test.ml
(** Testing different parts of code provided by professor **)
(** Also try different things with the graphics module to see how all the
    parts work together **)

(*Game state*)
(*I think this tracks the variable parts of the game like cursor, grid etc*)
module State = struct (*Custom module called State*)

  (*Initializes the cursor in bottom left corner*)
  let make (x,y) =
     (0, 0)

  (* Push cursor position by (fx, fy) unless it's at the window boundry*)
  let push (w,h) (fx, fy) state =
    let (x, y) = state in
      if (x + fx < 0) then      (0, y + fy)
      else if (x + fx >= w) then (x, y + fy)
      else if (y + fy < 0) then (x + fx, 0)
      else if (y + fy >= h) then (x + fx, y)
      else 
	(x + fx, y + fy)
end


(* event type to take user input *)
(* Exit = q 
   Select = j
   Cancel = k *)
type event = 
  Up | Down | Left | Right | Exit | Select | Cancel

let get_event () = 
  if Graphics.key_pressed () then begin
    let ch = Graphics.read_key () in
    match ch with
    | 'w' | 'W' -> Some Up
    | 'a' | 'A' -> Some Left
    | 'd' | 'D' -> Some Right
    | 's' | 'S' -> Some Down
    | 'q' | 'Q' -> Some Exit
    | 'j' | 'J' -> Some Select
    | 'k' | 'K' -> Some Cancel    
    | _ -> None
  end else
    None


(*Playing with unit types here*)
(*
  type unit =
  Melee | Range | Horse | Default
*)
(*Playing with unit module here*)
(*
module Unit = struct
  type  =
    { hp : int;
      atk : int;
      range : int;
      move : int;
    };;
end
*)


(* terrain types to draw map tiles *)
type terrain =
  Default | Forest | Hills

(* Board will be 10x10 for now *)
(*   -----------------------------------------
r9   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r8   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r7   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r6   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r5   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r4   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r3   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r2   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r1   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r0   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
      c0  c1  c2  c3  c4  c5  c6  c7  c8  c9   *)
let board =
  let d = Default in
  let f = Forest in
  let h = Hills in
(*r0*) [ [h; h; h; h; h; h; h; h; h; h];
(*r1*)   [h; d; d; d; h; f; f; f; f; h];
(*r2*)   [h; d; d; d; h; f; f; f; d; h];
(*r3*)   [h; d; d; d; h; f; d; d; f; h];
(*r4*)   [h; h; h; d; h; f; f; d; f; h];
(*r5*)   [h; f; f; d; f; f; f; d; f; h];
(*r6*)   [h; f; f; d; d; d; f; d; f; h];
(*r7*)   [h; f; f; f; d; d; d; d; f; h];
(*r8*)   [h; f; f; f; f; f; f; f; f; h];
(*r9*)   [h; h; h; h; h; h; h; h; h; h]; ]

let draw state = 
  (* BG gets drawn first and everything else overlayed on top*)
  (* Draw white background the size of the screen from bottom left *)
  Graphics.set_color Graphics.white;
  Graphics.fill_rect 0 0 (Graphics.size_x()) (Graphics.size_y());

  (* Draw game board here *)
  let unit = Graphics.size_x()/10 in
    let base_color terrain xpos ypos=
	let draw_thing thing =
	  "img/"^thing
	  |> Load_image.load_array
	  |> Load_image.sub_image 0 0 (unit) (unit) (Some (0x00FFFF))
	in

	if terrain = Default then (
	  Graphics.set_color 0x999900;
	  Graphics.fill_rect xpos ypos unit unit		  
	)
	else if terrain = Forest then 
	  Graphics.draw_image (draw_thing "tree.png") xpos ypos
	else if terrain = Hills then
	  Graphics.draw_image (draw_thing "mountain.png") xpos ypos
	  
	 (* Graphics.set_color 0x999999 *)
    in

    let rec draw_row xpos ypos = function
       | hd::tl -> 
	base_color hd xpos ypos; 
	Graphics.set_color Graphics.black;
	Graphics.set_line_width 1;
	Graphics.draw_rect xpos ypos unit unit;
	draw_row (xpos+unit) (ypos) tl
       | [] -> ()
    in
    
    let rec draw_board_aux ypos = function
      | hd::tl -> begin
		draw_row 0 ypos hd;
		draw_board_aux (ypos+unit) tl;
		end
      | [] -> ()
    in 

    let draw_board gameboard = draw_board_aux 0 gameboard; 
    in draw_board board;


 
  (* Testing unit drawing here *)
  let test_draw =
    "img/knight.png"
    |> Load_image.load_array
    |> Load_image.sub_image 0 0 (unit)  (unit) (Some (0x00FFFF))
  in
  Graphics.draw_image test_draw 100 100;

  (* Testing cpu unit drawing here *)
  let test_draw2 =
    "img/knight_bad.png"
    |> Load_image.load_array
    |> Load_image.sub_image 0 0 (unit)  (unit) (Some (0x00FFFF))
  in
  Graphics.draw_image test_draw2 700 100;
  Graphics.draw_image test_draw2 700 200;
  Graphics.draw_image test_draw2 800 100;





  (* Draw a cursor with a transparent middle so you can see the tile below *)
  let (x, y) = state in
  Graphics.set_color 0x7D8BCF;
  Graphics.set_line_width 3;
  Graphics.draw_rect x y ((Graphics.size_x()-2)/10) ((Graphics.size_y()-2)/10);
  Graphics.set_color Graphics.black;
  Graphics.draw_rect x y (Graphics.size_x()/10) (Graphics.size_y()/10);
  Graphics.synchronize()

let () =
  Graphics.open_graph " 1000x1000";
  (*This line added to solve screen flickering*)
  Graphics.auto_synchronize false;
  at_exit Graphics.close_graph;

  (* A square is 1/10 of the total size*)
  let unit = Graphics.size_x()/10 in
  let w = Graphics.size_x() in
  let h = Graphics.size_y() in
 

  let rec loop st =

    (*I should probably keep this sleep so I dont blow up my Pi lol*)
    Unix.sleep 2;

    match get_event () with
    | Some Exit -> ()
    | opt ->
	let st2 = (*I needed to add the %! because it wouldn't print otherwise :/ *)
	  match opt with
	  | None -> st
	  | Some Left -> Printf.printf "Left\n%!";
		State.push (w,h) (-unit,0) st
	  | Some Up -> Printf.printf "Up\n%!";
		State.push (w,h) (0,unit) st
	  | Some Right -> Printf.printf "Right\n%!";
		State.push (w,h) (unit,0) st
	  | Some Down -> Printf.printf "Down\n%!";
		State.push (w,h) (0,-unit) st
	  | Some Select -> Printf.printf "Select\n%!";
		st
	  | Some Cancel -> Printf.printf "Cancel\n%!";
		st
	  | Some _ -> st
	in

	(*draw*)
	draw st2;

	(*calling loop again*)
	loop st2
  in
  let w = Graphics.size_x () in
  let h = Graphics.size_y () in

  loop (State.make (w,h))


load_image.ml
(* Using the prof's example of image loading *)
(* he uses this function to convert the image to ints *)
let load_array filename = 
  let img = ImageLib.openfile filename in
  let w = img.Image.width in (*looks similar to graphics*)
  let h = img.Image.height in
  let max_val = img.Image.max_val in (*unsure what max val is*)

  Array.init h (fun y ->
    Array.init w (fun x -> 
      Image.read_rgb img x y
	(fun r g b ->
	  r *256 *256 *255 /max_val +
	  g *256 *255 /max_val +
	  b *255 /max_val)
    )
  )

let sub_image x y w h opt_transp arr =
  let sub_arr = 
    Array.init h (fun j ->
      let row = Array.sub arr.(y+j) x w in
      match opt_transp with
      | Some transpcolor -> Array.map (fun c -> if c = transpcolor then Graphics.transp else c) row
      | None -> row
    )
  in 
  Graphics.make_image sub_arr


draw_screen.ml
(* I want to separate all the draw functions here to clean up the main prog.
  It's getting messy and hard to follow lol *)

(* terrain types to draw map tiles *)
type terrain =
  Default | Forest | Hills

(* Board will be 10x10 for now *)
(*   -----------------------------------------
r9   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r8   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r7   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r6   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r5   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r4   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r3   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r2   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r1   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r0   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
      c0  c1  c2  c3  c4  c5  c6  c7  c8  c9   *)

let board =
  let d = Default in
  let f = Forest in
  let h = Hills in
(*r0*) [ [h; h; h; h; h; h; h; h; h; h];
(*r1*)   [h; d; d; d; h; f; f; f; f; h];
(*r2*)   [h; d; d; d; h; f; f; f; d; h];
(*r3*)   [h; d; d; d; h; f; d; d; f; h];
(*r4*)   [h; h; h; d; h; f; f; d; f; h];
(*r5*)   [h; f; f; d; f; f; f; d; f; h];
(*r6*)   [h; f; f; d; d; d; f; d; f; h];
(*r7*)   [h; f; f; f; d; d; d; d; f; h];
(*r8*)   [h; f; f; f; f; f; f; f; f; h];
(*r9*)   [h; h; h; h; h; h; h; h; h; h]; ]

(* I want this function to be in charge of calling all other draw
  funtions. 
  Order of operation:
  -board
  -units
  -cursor *)
(*let draw_all board units cursor =*)
  (*call function to draw board*)
    (*use <board_img> initialized at beginning of program*)
  (*call function to draw units*)
  (*call function to draw cursor*)



(* Function to initialize game board. Returns Graphics type <image> *)
let board_init tile_list =
  (* Draw game board here *)
  let unit = Graphics.size_x()/10 in
  let base_color terrain xpos ypos=
      let draw_thing thing =
        "img/"^thing
        |> Load_image.load_array
        |> Load_image.sub_image 0 0 (unit) (unit) (Some (0x00FFFF))
      in

      if terrain = Default then (
        Graphics.set_color 0x999900;
        Graphics.fill_rect xpos ypos unit unit
      )
      else if terrain = Forest then
        Graphics.draw_image (draw_thing "tree.png") xpos ypos
      else if terrain = Hills then
        Graphics.draw_image (draw_thing "mountain.png") xpos ypos
  in

  let rec draw_row xpos ypos = function
     | hd::tl ->
      base_color hd xpos ypos;
      Graphics.set_color Graphics.black;
      Graphics.set_line_width 1;
      Graphics.draw_rect xpos ypos unit unit;
      draw_row (xpos+unit) (ypos) tl
     | [] -> ()
  in

  let rec draw_board_aux ypos = function
    | hd::tl -> begin
              draw_row 0 ypos hd;
              draw_board_aux (ypos+unit) tl;
              end
    | [] -> ()
  in

  let draw_board gameboard = draw_board_aux 0 gameboard;
  in draw_board tile_list;

  (* This captures everything that's been drawn on the screen and returns
     as type image *)
  Graphics.get_image 0 0 (Graphics.size_x()) (Graphics.size_y())



g_test_orig.ml
(** different things with the graphics module to see how all the
    parts work together **)

(*Game state*)
(*I think this tracks the variable parts of the game like cursor, grid etc*)
module State = struct (*Custom module called State*)

  (*Initializes the cursor in bottom left corner*)
  let make (x,y) =
     (0, 0)

  (* Push cursor position by (fx, fy) unless it's at the window boundry*)
  let push (w,h) (fx, fy) state =
    let (x, y) = state in
      if (x + fx < 0) then      (0, y + fy)
      else if (x + fx >= w) then (x, y + fy)
      else if (y + fy < 0) then (x + fx, 0)
      else if (y + fy >= h) then (x + fx, y)
      else 
	(x + fx, y + fy)
end


(* event type to take user input *)
(* Exit = q 
   Select = j
   Cancel = k *)
type event = 
  Up | Down | Left | Right | Exit | Select | Cancel

let get_event () = 
  if Graphics.key_pressed () then begin
    let ch = Graphics.read_key () in
    match ch with
    | 'w' | 'W' -> Some Up
    | 'a' | 'A' -> Some Left
    | 'd' | 'D' -> Some Right
    | 's' | 'S' -> Some Down
    | 'q' | 'Q' -> Some Exit
    | 'j' | 'J' -> Some Select
    | 'k' | 'K' -> Some Cancel    
    | _ -> None
  end else
    None


(*Playing with unit types here*)
(*
  type unit =
  Melee | Range | Horse | Default
*)
(*Playing with unit module here*)
(*
module Unit = struct
  type  =
    { hp : int;
      atk : int;
      range : int;
      move : int;
    };;
end
*)


(* terrain types to draw map tiles *)
type terrain =
  Default | Forest | Hills

(* Board will be 10x10 for now *)
(*   -----------------------------------------
r9   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r8   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r7   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r6   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r5   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r4   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r3   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r2   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r1   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
r0   | D | F | H | D | F | H | D | F | H | D |
     -----------------------------------------
      c0  c1  c2  c3  c4  c5  c6  c7  c8  c9   *)
let board =
  let d = Default in
  let f = Forest in
  let h = Hills in
(*r0*) [ [h; h; h; h; h; h; h; h; h; h];
(*r1*)   [h; d; d; d; h; f; f; f; f; h];
(*r2*)   [h; d; d; d; h; f; f; f; d; h];
(*r3*)   [h; d; d; d; h; f; d; d; f; h];
(*r4*)   [h; h; h; d; h; f; f; d; f; h];
(*r5*)   [h; f; f; d; f; f; f; d; f; h];
(*r6*)   [h; f; f; d; d; d; f; d; f; h];
(*r7*)   [h; f; f; f; d; d; d; d; f; h];
(*r8*)   [h; f; f; f; f; f; f; f; f; h];
(*r9*)   [h; h; h; h; h; h; h; h; h; h]; ]

let draw state = 
  (* BG gets drawn first and everything else overlayed on top*)
  (* Draw white background the size of the screen from bottom left *)
  Graphics.set_color Graphics.white;
  Graphics.fill_rect 0 0 (Graphics.size_x()) (Graphics.size_y());

  (* Draw game board here *)
  let unit = Graphics.size_x()/10 in
    let base_color terrain xpos ypos=
	let draw_thing thing =
	  "img/"^thing
	  |> Load_image.load_array
	  |> Load_image.sub_image 0 0 (unit) (unit) (Some (0x00FFFF))
	in

	if terrain = Default then (
	  Graphics.set_color 0x999900;
	  Graphics.fill_rect xpos ypos unit unit		  
	)
	else if terrain = Forest then 
	  Graphics.draw_image (draw_thing "tree.png") xpos ypos
	else if terrain = Hills then
	  Graphics.draw_image (draw_thing "mountain.png") xpos ypos
	  
	 (* Graphics.set_color 0x999999 *)
    in

    let rec draw_row xpos ypos = function
       | hd::tl -> 
	base_color hd xpos ypos; 
	Graphics.set_color Graphics.black;
	Graphics.set_line_width 1;
	Graphics.draw_rect xpos ypos unit unit;
	draw_row (xpos+unit) (ypos) tl
       | [] -> ()
    in
    
    let rec draw_board_aux ypos = function
      | hd::tl -> begin
		draw_row 0 ypos hd;
		draw_board_aux (ypos+unit) tl;
		end
      | [] -> ()
    in 

    let draw_board gameboard = draw_board_aux 0 gameboard; 
    in draw_board board;


 
  (* Testing unit drawing here *)
  let test_draw =
    "img/knight.png"
    |> Load_image.load_array
    |> Load_image.sub_image 0 0 (unit)  (unit) (Some (0x00FFFF))
  in
  Graphics.draw_image test_draw 100 100;

  (* Testing cpu unit drawing here *)
  let test_draw2 =
    "img/knight_bad.png"
    |> Load_image.load_array
    |> Load_image.sub_image 0 0 (unit)  (unit) (Some (0x00FFFF))
  in
  Graphics.draw_image test_draw2 700 100;
  Graphics.draw_image test_draw2 700 200;
  Graphics.draw_image test_draw2 800 100;





  (* Draw a cursor with a transparent middle so you can see the tile below *)
  let (x, y) = state in
  Graphics.set_color 0x7D8BCF;
  Graphics.set_line_width 3;
  Graphics.draw_rect x y ((Graphics.size_x()-2)/10) ((Graphics.size_y()-2)/10);
  Graphics.set_color Graphics.black;
  Graphics.draw_rect x y (Graphics.size_x()/10) (Graphics.size_y()/10);
  Graphics.synchronize()

let () =
  Graphics.open_graph " 1000x1000";
  (*This line added to solve screen flickering*)
  Graphics.auto_synchronize false;
  at_exit Graphics.close_graph;

  (* A square is 1/10 of the total size*)
  let unit = Graphics.size_x()/10 in
  let w = Graphics.size_x() in
  let h = Graphics.size_y() in
 

  let rec loop st =

    (*I should probably keep this sleep so I dont blow up my Pi lol*)
    Unix.sleep 2;

    match get_event () with
    | Some Exit -> ()
    | opt ->
	let st2 = (*I needed to add the %! because it wouldn't print otherwise :/ *)
	  match opt with
	  | None -> st
	  | Some Left -> Printf.printf "Left\n%!";
		State.push (w,h) (-unit,0) st
	  | Some Up -> Printf.printf "Up\n%!";
		State.push (w,h) (0,unit) st
	  | Some Right -> Printf.printf "Right\n%!";
		State.push (w,h) (unit,0) st
	  | Some Down -> Printf.printf "Down\n%!";
		State.push (w,h) (0,-unit) st
	  | Some Select -> Printf.printf "Select\n%!";
		st
	  | Some Cancel -> Printf.printf "Cancel\n%!";
		st
	  | Some _ -> st
	in

	(*draw*)
	draw st2;

	(*calling loop again*)
	loop st2
  in
  let w = Graphics.size_x () in
  let h = Graphics.size_y () in

  loop (State.make (w,h))
